Source/DafnyCore/Compilers/Java/Compiler-java.cs

```diff
@@ -3646,22 +3665,31 @@ protected class ClassWriter : IClassWriter {
       return type == NativeObjectType || type.IsTypeParameter;
     }
 
+    private bool IsCharBasedType(Type type) {
+      return type switch {
+        { IsCharType: true } => true,
+        UserDefinedType { ResolvedClass: SubsetTypeDecl std } =>
+          IsCharBasedType(std.RhsWithArgument(type.TypeArgs)),
+        _ => false
+      };
+    }
+
     protected override ConcreteSyntaxTree EmitCoercionIfNecessary(Type/*?*/ from, Type/*?*/ to, IToken tok, ConcreteSyntaxTree wr) {
       from = from == null ? null : DatatypeWrapperEraser.SimplifyType(from);
       to = to == null ? null : DatatypeWrapperEraser.SimplifyType(to);
 
       if (UnicodeCharEnabled) {
         // Need to box from int to CodePoint, or unbox from CodePoint to int
 
-        if (IsObjectType(from) && to is { IsCharType: true }) {
-          wr.Write($"((dafny.CodePoint)(");
+        if (IsObjectType(from) && IsCharBasedType(to)) {
+          wr.Write("((dafny.CodePoint)(");
           var w = wr.Fork();
           wr.Write(")).value()");
           return w;
         }
 
-        if (from is { IsCharType: true } && IsObjectType(to)) {
-          wr.Write($"dafny.CodePoint.valueOf(");
+        if (IsCharBasedType(from) && IsObjectType(to)) {
```

Comment:
`IsObjectType` is not part of this PR, but I find its name misleading. It seems to return whether or not the given type is _compiled into_ Java's type `java.lang.Object`. Could the method be renamed to something like `IsObjectTargetType` or something like that?

Reply:
I made it `IsNativeObjectType` which is an improvement, but I still don't love it. It's checking whether the type is `java.lang.Object` or a type parameter, and is meant to test whether boxing/unboxing is needed.